sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallInsecureSecurityAlgorithm.cs

```diff
@@ -0,0 +1,117 @@
+
+        protected sealed override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c => CheckObjectCreation(c),
+                SyntaxKind.ObjectCreationExpression);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c => CheckInvocation(c),
+                SyntaxKind.InvocationExpression);
+        }
+
+        private void CheckInvocation(SyntaxNodeAnalysisContext context)
+        {
+            var invocation = (InvocationExpressionSyntax)context.Node;
+
+            var methodSymbol = context.SemanticModel.GetSymbolInfo(invocation.Expression).Symbol as IMethodSymbol;
+            if (methodSymbol == null)
+            {
+                return;
+            }
+
+            if (methodSymbol.ReturnType.DerivesFromAny(AlgorithmTypes) ||
+                IsInsecureBaseAlgorithmCreationFactoryCall(methodSymbol, invocation.ArgumentList))
+            {
+                context.ReportDiagnostic(Diagnostic.Create(SupportedDiagnostics[0],
+                    invocation.GetLocation()));
+            }
+        }
+
+        private bool IsInsecureBaseAlgorithmCreationFactoryCall(IMethodSymbol methodSymbol,
+            ArgumentListSyntax argumentList)
+        {
+            if (argumentList == null ||
+                methodSymbol?.ContainingType == null ||
+                methodSymbol.Name == null)
+            {
+                return false;
+            }
+
+            var methodFullName = $"{methodSymbol.ContainingType}.{methodSymbol.Name}";
+
+            if (argumentList.Arguments.Count == 0)
+            {
+                return AlgorithmParameterlessFactoryMethods.Contains(methodFullName);
+            }
+
+            if (argumentList.Arguments.Count > 1 ||
+                !argumentList.Arguments.First().Expression.IsKind(SyntaxKind.StringLiteralExpression))
+            {
+                return false;
+            }
+
+            if (!AlgorithmParameteredFactoryMethods.Contains(methodFullName))
+            {
+                return false;
+            }
+
+            var literalExpressionSyntax = (LiteralExpressionSyntax)argumentList.Arguments.First().Expression;
+            return FactoryParameterNames.Any(alg => alg.Equals(literalExpressionSyntax.Token.ValueText, StringComparison.Ordinal));
+        }
+
+        private void CheckObjectCreation(SyntaxNodeAnalysisContext context)
+        {
+            var objectCreation = (ObjectCreationExpressionSyntax)context.Node;
+
+            var typeInfo = context.SemanticModel.GetTypeInfo(objectCreation);
+            if (typeInfo.ConvertedType == null || typeInfo.ConvertedType is IErrorTypeSymbol)
+            {
+                return;
+            }
+
+            if (typeInfo.ConvertedType.DerivesFromAny(AlgorithmTypes))
+            {
+                context.ReportDiagnostic(Diagnostic.Create(SupportedDiagnostics[0],
```

Comment:
I really don't like SupportedDiagnostics[0], may be we should come with a nice way to handle the situation in base classes. Perhaps create one (or more) abstract properties that will return the diagnostics, so that we don't end with:
SupportedDiagnostics[0], 1, 2, etc... in too many places.